core/src/main/java/io/grpc/internal/ClientStream.java

```diff
@@ -72,6 +73,7 @@
    * <p>This method should not throw any exceptions.
    *
    * @param listener non-{@code null} listener of stream events
+   * @param headers to send at the beginning of the call
    */
-  void start(ClientStreamListener listener);
+  void start(ClientStreamListener listener, Metadata headers);
```

Comment:
Yes, I had been planning to do this at the end of this pull request.

I was trying to removing the the Metadata from ClientTransport.newStream(), but got some issues with InProcessTransport and CallCredentialsApplyingTransportFactory. A bunch of changes seems needed to make the new newStream API work. Do you think we should address that in this pull request or are you fine with opening a new issue and solve it in a separate PR?

Not sure if I follow, why do you think ClientCallImpl is broken right now?